### ISSUE TITLE
Swizzle ProtoFlux Nodes

### ISSUE THREAD

**Issue by lxw404** [Yellow-Dog-Man/Resonite-Issues#1448]
### Is your feature request related to a problem? Please describe.

Currently, when extracting individual components from vector inputs/outputs or reordering them, the only solution is by using the `Unpack<T>` / `Pack<T>` nodes, however often it is the case that only a single value is required, or that the user simply wants to swap/reorder the components of one vector to another.

### Describe the solution you'd like

There should be an additional way to extract and reorder individual elements from vectors (swizzle) and a way to specify the exact elements to reorder.

This should allow for example a `float3` to `float2` swizzle node to take individual `float` elements from the original `float3` (`<X, Y, Z>`) and map them to a new `float2` in any order, such as `<Z, X>`.

The swizzle should also be completely arbitrary allowing you to map for example, a `float2` (`<X, Y>`) to `float4` as `<Y, Y, Y, Y>` which is often necessary for vector computations.

This is a common pattern in many scenarios which currently requires two nodes to accomplish and additionally, a very limited subset of (not all possible) swizzle drivers currently exists as components.

### Describe alternatives you've considered

The alternative is to use the `Unpack<T>` and `Pack<T>` nodes.

### Requesters

LuxKitty

**Comment by XDelta** [comment]
I don't think a swizzle type node would really be needed as each one of these conversions, remapping, and arbitrary calculations or casts in between are already possible with the more flexible `Unpack`/`Pack` nodes. Is there some specific problem you are encountering using these 2 that a swizzle node would solve beyond just having 1 less node?

![cf7c6e40d0](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/7883807/dc9bd9b9-79f0-469d-b1ae-6b506d3a291f)

**Comment by lxw404** [comment]
Swizzle is a basic function of [GLSL](https://www.khronos.org/opengl/wiki/Data_Type_(GLSL)#Swizzling), various other languages, and even other [visual node based languages](https://helpx.adobe.com/substance-3d-designer/function-graphs/nodes-reference-for-function-graphs/vector-and-swizzle-nodes.html) which reduces clutter when referring to portions of vectors, which I would assume would be a general good idea to follow (especially when we eventually have ProtoFlux-based shading nodes, this would be a *very* common design pattern).

Pack/Unpack is fine for the general case of doing small numbers of such conversions.  When you scale up to many many vectors requiring the same swizzle, the pack/unpack becomes redundant and very messy to follow.  Further, if you scale this up to N dimensional vectors, you have an even bigger problem of tracing the wires between the pack/unpack.

**Comment by Frooxius** [comment]
We can add these, but the problem here will be the workflow. I don't think adding these before we have search and filtering would help that much, because you'd have to scroll through hundreds of variants to find the swizzle you want.

**Comment by lxw404** [comment]
The way that Substance handles this specifically is by providing an overloaded data type as a default, which reduces the possible combinations to just 4 swizzle output types, and then allowing the user to use an enum selection per element on the node itself.

<p align="middle">
  <img align="middle" src="https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/52231149/5f7b14ca-0129-4c6a-975f-8907c9ab51a9" width="30%" />
  <img align="middle" src="https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/52231149/aa0c39bd-3221-4f49-99f3-eaa784241e62" width="30%" /> 
  <img align="middle" src="https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/52231149/085fa151-8a9d-4bd4-be7b-33d32b78538f" width="30%" />
</p>

The workflow becomes extremely intuitive from this as you just have to focus on:
1. What is the current desired output vector size
2. Next, which swizzle combination you like by changing the enums
3. Then finally plug in an input to change the overloaded type to your desired input

In the example shown below, I chose a `Float2` output size, then made it a `<Y, Y>` swizzle by choosing the enum for Y on both vector elements, then finally you can plug in any input vector and it will change its overloaded type.  So for example if I put a `Float4` into the input, it now solidifies its final type to be `Float4` to `Float2`, with a swizzle of `<Y, Y>`.

I would imagine listing every possible combination of every single swizzle would be absolutely nightmarish for an end user, even with a search function.

**Comment by Frooxius** [comment]
It's interesting how they approach it, but we don't have that kind of mechanism and I'm not sure if it would fit well with our system, we'll have to see how to approach it. The point here is - I don't think we can add this without additional system to make the selection easy.

I don't think the search would be too bad for this - you simply just type the swizzle you want, like "xxyz" and you get the node. I'm not saying this is the way we go with this, we might build nicer UI for that, but I don't see that being as nightmarish.

**Comment by lxw404** [comment]
I suppose that does expose a limitation to the current system.  With that specific node based programming system, there is the concept of *node properties* which are arbitrary data stored within the node itself that can either be internal, or create UI for the end user to manipulate.

I am curious on the search mechanism for this and how it could be not as cluttered, as currently it is a little cumbersome to pick exact node types by typing their full names in.  Also would it be possible to just use `Dummy` as a default input type to start with to reduce the total combinations?

**Comment by Frooxius** [comment]
It's not a limitation, but different design, which requires different approach. ProtoFlux is designed with high performance execution in mind. We could add node properties (in fact, some nodes do use this system), but swizzles should be really fast operation and checking the type of swizzle on every evaluation is excessive.

We're getting a bit off topic here. We have issue for the search and such here: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/578

**Comment by lxw404** [comment]
Thank you for linking that.  I would definitely hope that it wouldn't do type checking every evaluation for sure, but rather that a UI selection would then choose the appropriate static type internally or something of that sort.  For Substance specifically this amounts to compiling the subnode graph into its final fragment shader representation, which happens on first evaluation.

**Comment by Frooxius** [comment]
Yes, however ProtoFlux is structured somewhat differently, so we have to structure and build this around its mechanism - which is what I am proposing here.
